server/src/test/java/org/elasticsearch/action/support/CountDownActionListenerTests.java

```diff
@@ -0,0 +1,190 @@
+            threads[i].start();
+        }
+        for (Thread t : threads) {
+            t.join();
+        }
+        assertTrue(called.get());
+    }
+
+    public void testFailed() {
+        AtomicBoolean called = new AtomicBoolean(false);
+        AtomicReference<Exception> excRef = new AtomicReference<>();
+
+        ActionListener<Void> result = new ActionListener<>() {
+            @Override
+            public void onResponse(Void ignored) {
+                called.set(true);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                excRef.set(e);
+            }
+        };
+        int size = randomIntBetween(3, 4);
+        CountDownActionListener listener = new CountDownActionListener(size, result);
+        listener.onResponse(null);
+        IOException ioException = new IOException();
+        RuntimeException rtException = new RuntimeException();
+        listener.onFailure(rtException);
+        listener.onFailure(ioException);
+        if (size == 4) {
+            listener.onResponse(null);
+        }
+        assertNotNull(excRef.get());
+        assertEquals(rtException, excRef.get());
+        assertEquals(1, excRef.get().getSuppressed().length);
+        assertEquals(ioException, excRef.get().getSuppressed()[0]);
+        assertFalse(called.get());
+    }
+
+    public void testValidation() throws InterruptedException {
+        AtomicBoolean called = new AtomicBoolean(false);
+        ActionListener<Void> result = new ActionListener<>() {
+            @Override
+            public void onResponse(Void ignored) {
+                called.compareAndSet(false, true);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                called.compareAndSet(false, true);
+            }
+        };
+
+        // can't use a groupSize of 0
+        expectThrows(AssertionError.class, () -> new CountDownActionListener(0, result));
+
+        // can't use a null listener or runnable
+        expectThrows(NullPointerException.class, () -> new CountDownActionListener(1, (ActionListener<Void>) null));
+        expectThrows(NullPointerException.class, () -> new CountDownActionListener(1, (Runnable) null));
+
+        final int overage = randomIntBetween(1, 10);
+        AtomicInteger assertionsTriggered = new AtomicInteger();
+        final int groupSize = randomIntBetween(10, 1000);
+        AtomicInteger count = new AtomicInteger();
+        CountDownActionListener listener = new CountDownActionListener(groupSize, result);
+        int numThreads = randomIntBetween(2, 5);
+        Thread[] threads = new Thread[numThreads];
+        CyclicBarrier barrier = new CyclicBarrier(numThreads);
+        for (int i = 0; i < numThreads; i++) {
+            threads[i] = new Thread(() -> {
+                try {
+                    barrier.await(10, TimeUnit.SECONDS);
+                } catch (Exception e) {
+                    throw new AssertionError(e);
```

Comment:
I think this will disappear into the void and not be reported...

Reply:
It should be caught by an uncaught exception handler and result in a test failure

```
Dec 19, 2022 6:34:54 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[Thread-2,5,TGRP-CountDownActionListenerTests]
java.lang.AssertionError: boom
	at __randomizedtesting.SeedInfo.seed([39A5DC05E92E3853]:0)
	at org.elasticsearch.action.support.CountDownActionListenerTests.lambda$testValidation$4(CountDownActionListenerTests.java:136)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Otherwise, yikes, we'd have a lot of tests that aren't asserting what we think they are!

Reply:
Great!

Reply:
They get caught and rethrown as a `com.carrotsearch.randomizedtesting.UncaughtExceptionError` by the test runner, and they do get logged. I guess that's not quite as obvious as directly failing the test, but `Thread` and `CyclicBarrier` aren't the class under tests either (and not where I'd expect issues to come from here).

Reply:
`:slowpoke:` <-- me 😄